### ISSUE TITLE
Docker backend without docker.domain generates bad host rules

### ISSUE THREAD

**Issue by cpuguy83** [containous/traefik#1317]
### What version of Traefik are you using (`traefik version`)?
1.1.2

### What is your environment & configuration (arguments, toml...)?

```
--debug --web --docker --acme --acme.email=cpuguy83@gmail.com --acme.ondemand=true --acme.storage=/var/acme/acme.json --acme.entrypoint=https --accesslogsfile=/var/acme/access.log --acme.onhostrule=true --entryPoints='Name:https Address::443 TLS' --entryPoints='Name:http Address::80' --acme.caserver="https://acme-v01.api.letsencrypt.org/directory"
```

### What did you do?

Deploy a global service (specifically service for docker-cloud proxy).

### What did you expect to see?

Traefik should not be requesting certs for this service

### What did you see instead?

Traefik is requesting certs for the service even though there's no traefik related labels on it.
This causes all acme requests to fail.

Even getting traefik+acme to work at all requires tons of labels with the services I do want certs and routing for, e.g.:

```
      - "traefik.frontend.rule=Host:container42.com,www.container42.com"
      - "traefik.frontend.entryPoints=http,https"
      - "traefik.protocol=http"
      - "traefik.port=80"
```

But just deploying a service without any labels, traefik seems to pick it up and try to request certs for it.
It seems like it definitely shouldn't request certs for services without a `traefik.frontend.rule` label since there's no way for it to know what domain to request certs for.

**Comment by vdemeester** [comment]
@cpuguy83 you are using the docker provider without swarm mode right ? In that case, if a container has no label and only one port exposed, it will use the default domain with the name of the container if I remember correctly. 

One way to disable that is to set `exposedByDefault` to `false` (but this make you add one more label 😅 👼)

```
# Expose containers by default in traefik
# If set to false, containers that don't have `traefik.enable=true` will be ignored
#
# Optional
# Default: true
#
exposedbydefault = true
```

But looking at `acme.onhostrule`, I feel there is a bug 😓

**Comment by cpuguy83** [comment]
With swarm mode.

**Comment by vdemeester** [comment]
@cpuguy83 I meant without `--docker.swarmmode` for træfik 👼

**Comment by cpuguy83** [comment]
Oh, I didn't know that was a thing!

**Comment by emilevauge** [comment]
In theory, Traefik is **not** requesting for an ACME cert without any domain guessed from rules (based on our code https://github.com/containous/traefik/blob/v1.2/rules.go#L198).
But even without label, you have a default configuration that is used by Traefik to assign a domain name to a container: `{containerName}.domain` (from https://github.com/containous/traefik/blob/v1.2/traefik.sample.toml#L395).
If you don't need a cert, you can remove your container from your `https` entrypoint with label: `traefik.frontend.entryPoints=http`.
@cpuguy83 does that fix your issue ?

**Comment by cpuguy83** [comment]
@emilevauge It's not a great solution, in particular because the services I do want acme certs for (or perhaps just to have them served as HTTPS) require me to add the https entrypoint.

There's a lot of weird behavior that seems to happen when acme is enabled.

**Comment by emilevauge** [comment]
@cpuguy83 I'm not sure to understand what's wrong with this solution. You can have both services with AND without https (+ACME), you can have a default configuration (http, https, http+https) and override it for each service. I don't know how it could be better.

>There's a lot of weird behavior that seems to happen when acme is enabled.

Could you give more details on this ? Would love to help :)

**Comment by emilevauge** [comment]
@cpuguy83 

>Added to container labels:
>
>```
>- "traefik.frontend.entryPoints=http,https"
>```

To avoid adding this on each container, you can add `--defaultentrypoints="http,https"` to your command line.


>And I also have disable traefik on backends I don't wanted proxied via label otherwise I get acme requests for them:
>
>```
>- "traefik.enable=false"
>```

Either Traefik exposes by default all containers and you have to disable them manually, either you can should to **not** expose by default your containers, and enable them manually (adding `--docker.exposedbydefault` to your command line).

>At this point there's some weird mixture of settings that are required to make things work at all, let alone the acme stuff.

You can already choose different default behaviors, what could be improved here according to you?

>At the same time it's really weird to request certs for services without a host rule, especially since this seems to break all other cert requests, even more strange if you haven't explicitly set a domain since the host+domain are going to be nonsense.

There is an issue if you don't explicitly set a `docker.domain` indeed. And the issue is not linked to ACME but with Docker configuration backend. It appears that an empty `docker.domain` ends up in a `Host:{container_name}.` rule while this container should be filtered.

**Comment by cpuguy83** [comment]
@emilevauge Why do default (entrypoints) change when enabling acme?

**Comment by emilevauge** [comment]
@cpuguy83 they don't. But ACME needs an entrypoint reference (`--acme.entrypoint`).

**Comment by emilevauge** [comment]
Fixed by #1401
